sklearn/cluster/hierarchical.py

```diff
@@ -444,6 +447,39 @@ def linkage_tree(X, connectivity=None, n_components=None,
         assert n_clusters <= n_samples
         n_nodes = 2 * n_samples - n_clusters
 
+    if linkage == 'single':
+        from scipy.sparse.csgraph import minimum_spanning_tree
+
+        # Use scipy.sparse.csgraph to generate a minimum spanning tree
+        mst = minimum_spanning_tree(connectivity.tocsr())
+
+        # Convert the graph to scipy cluster array format
+        nonzeros = mst.nonzero()
+        nonzero_vals = mst.data
+        mst_array = np.vstack(nonzeros + (nonzero_vals,)).T
+
+        # Sort edges of the min_spanning_tree by weight
+        mst_array = mst_array[np.argsort(mst_array.T[2]),:]
+
+        # Convert edge list into standard hierarchical clustering format
+        single_linkage_tree = _hierarchical.single_linkage_label(mst_array)
+        children_ = single_linkage_tree[:, :2].astype(np.int)
+
+        # Compute parents
+        parent = np.arange(n_nodes, dtype=np.intp)
+        for i, (left, right) in enumerate(children_, n_samples):
```

Comment:
Here is a scaling performance comparison for the sparse matrix case:
![image](https://user-images.githubusercontent.com/11962885/28248580-9260824e-6a0c-11e7-93fa-70b2305eed56.png)
Most of the time is single linkage is actually spent fixing the connectivity matrix -- I'm factoring that out of the test and will post that soon, but it already looks pretty good.

The dense case is simply scipy.cluster.hierarchy, so I presume the performance of single linkage there is already demonstrated.